torchdynamo: add convolution pointwise(binary) fusion kernel

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #86583
* __->__ #86582
* #86581

Support binary fusion of Convolution with:

- add
- sub
- mul
- div

XiaobingSuper committed:
torchdynamo: add convolution binary fusion kernel

[ghstack-poisoned]

XiaobingSuper committed:
Update on "torchdynamo: add convolution binary fusion kernel"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "torchdynamo: add convolution pointwise(binary) fusion kernel"


Support binary fusion of Convolution with:

- add
- sub
- mul
- div

[ghstack-poisoned]

XiaobingSuper committed:
Update on "torchdynamo: add convolution pointwise(binary) fusion kernel"


Support binary fusion of Convolution with:

- add
- sub
- mul
- div

[ghstack-poisoned]

EikanWang commented:
@jgong5 , any comments for this PR?

jgong5 commented on test/test_mkldnn_fusion.py:
```diff
@@ -246,5 +255,40 @@ def forward(self, x):
+                    oC = 10 * groups
+                    iC = 3 * groups
+                    x_shape = (1, iC) + input_shapes[dim]
+                    x = torch.randn(x_shape, dtype=torch.float32).to(memory_format=memory_format)
```
Test bfloat16 too? Same comments for other tests.

XiaobingSuper replied:
I think we can add the bfloat16 test case at another PR which includes linear, Conv.

jgong5 commented on test/test_mkldnn_fusion.py:
```diff
@@ -246,5 +255,40 @@ def forward(self, x):
+                    x = torch.randn(x_shape, dtype=torch.float32).to(memory_format=memory_format)
+                    mod = M(pointwise_fn, dim, iC, oC, dilation, groups, bias, kernel_size=3)
+                    mod = mod.to(memory_format=memory_format).eval()
+                    other = torch.randn_like(mod.conv(x))
```
Test same dtype and different dtype?

XiaobingSuper replied:
Now, I add a dtype check which makes sure all input should have same type. and for mixed-dtype, we will report an error to user.

XiaobingSuper committed:
Update on "torchdynamo: add convolution pointwise(binary) fusion kernel"


Support binary fusion of Convolution with:

- add
- sub
- mul
- div

[ghstack-poisoned]

XiaobingSuper commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -296,6 +296,116 @@ Tensor mkldnn_convolution_pointwise(
+      output_sizes == other_t.sizes(),
+      "Binary Fusion's add should have same shape");
+  bool can_be_fused = input_t.scalar_type() == other_t.scalar_type() &&
+      input_t.device() == other_t.device() && groups == 1 &&
```
Yes, add device check.

EikanWang commented:
@jansel , could you help review this PR for TorchInductor?

EikanWang commented:
@XiaobingSuper , please fix the CI issue.

XiaobingSuper commented:
> @XiaobingSuper , please fix the CI issue.

The error is not related to this PR.

XiaobingSuper committed:
Update on "torchdynamo: add convolution pointwise(binary) fusion kernel"


Support binary fusion of Convolution with:

- add
- sub
- mul
- div

[ghstack-poisoned]

XiaobingSuper committed:
Update on "torchdynamo: add convolution pointwise(binary) fusion kernel"


Support binary fusion of Convolution with:

- add
- sub
- mul
- div

[ghstack-poisoned]

EikanWang commented:
@pytorchbot merge
